# Replicate current behaviour for closed shells
Once that works we can move on to checking that open shells work.

* [ ] Use unrestricted method for closed shell

* [ ] Check that values match restricted, incl energy contributions

* [ ] Try with open shell

* [ ] Change driver back to allow choice of restricted scf

* [ ] Insert new functions

* [ ] Update pyscf localizers to do unrestricted


Need to force driver to always use unrestricted method.

In [3]:
from nbed.driver import NbedDriver
from pathlib import Path

args = {
    "geometry": str(Path("molecular_structures/cyclopentane.xyz").absolute()),
    "basis": "STO-3G",
    "xc_functional": "b3lyp",
    "n_active_atoms":2,
    "projector": "mu",
    "localization": "spade",
    "convergence": 1e-6,
    "savefile": None,
    "run_ccsd_emb": True,
    "run_fci_emb": False,
    'charge':0,
    'spin':0,
}

## Run restricted to get componenets

In [6]:
driver = NbedDriver(**args)

In [7]:
import logging
logger = logging.getLogger(__name__)
from nbed.localizers import Localizer
from typing import Optional, Tuple, Dict
import numpy as np
from pyscf import scf, dft, gto
from scipy import linalg
from nbed.exceptions import NbedConfigError

class OldSPADELocalizer(Localizer):
    """Localizer Class to carry out SPADE"""

    def __init__(
        self,
        pyscf_scf: gto.Mole,
        n_active_atoms: int,
        occ_cutoff: Optional[float] = 0.95,
        virt_cutoff: Optional[float] = 0.95,
        run_virtual_localization: Optional[bool] = False,
    ):
        super().__init__(
            pyscf_scf,
            n_active_atoms,
            occ_cutoff=occ_cutoff,
            virt_cutoff=virt_cutoff,
            run_virtual_localization=run_virtual_localization,
        )

    def _localize(
        self,
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """Localise orbitals using SPADE.
        Returns:
            active_MO_inds (np.array): 1D array of active occupied MO indices
            enviro_MO_inds (np.array): 1D array of environment occupied MO indices
            c_active (np.array): C matrix of localized occupied active MOs (columns define MOs)
            c_enviro (np.array): C matrix of localized occupied ennironment MOs
            c_loc_occ (np.array): full C matrix of localized occupied MOs
        """
        logger.info("Localising with SPADE.")
        n_occupied_orbitals = np.count_nonzero(self._global_ks.mo_occ)

        occupied_orbitals = self._global_ks.mo_coeff[:, :n_occupied_orbitals]

        n_act_aos = self._global_ks.mol.aoslice_by_atom()[self._n_active_atoms - 1][-1]
        logger.debug(f"{n_act_aos} active AOs.")

        ao_overlap = self._global_ks.get_ovlp()

        # Orbital rotation and partition into subsystems A and B
        # rotation_matrix, sigma = embed.orbital_rotation(occupied_orbitals,
        #    n_act_aos, ao_overlap)

        rotated_orbitals = (
            linalg.fractional_matrix_power(ao_overlap, 0.5) @ occupied_orbitals
        )
        _, sigma, right_vectors = linalg.svd(rotated_orbitals[:n_act_aos, :])

        logger.debug(f"Singular Values: {sigma}")

        # n_act_mos, n_env_mos = embed.orbital_partition(sigma)
        # Prevents an error with argmax
        if len(sigma) == 1:
            n_act_mos = 1
        else:
            value_diffs = sigma[:-1] - sigma[1:]
            n_act_mos = np.argmax(value_diffs) + 1
        n_env_mos = n_occupied_orbitals - n_act_mos
        logger.debug(f"{n_act_mos} active MOs.")
        logger.debug(f"{n_env_mos} environment MOs.")

        # get active and enviro indices
        active_MO_inds = np.arange(n_act_mos)
        enviro_MO_inds = np.arange(n_act_mos, n_act_mos + n_env_mos)

        # Defining active and environment orbitals and density
        c_active = occupied_orbitals @ right_vectors.T[:, :n_act_mos]
        c_enviro = occupied_orbitals @ right_vectors.T[:, n_act_mos:]
        c_loc_occ = occupied_orbitals @ right_vectors.T

        # storing condition used to select env system
        self.enviro_selection_condition = sigma

        alpha = active_MO_inds, enviro_MO_inds, c_active, c_enviro, c_loc_occ
        beta = None

        return (alpha, beta)

    def run(self, sanity_check: bool = False) -> None:
        """Function that runs localization.

        Args:
            sanity_check (bool): optional flag to check denisty matrices and electron number after orbital localization
                                 makes sense
        """
        alpha, beta = self._localize()

        (
            self.active_MO_inds,
            self.enviro_MO_inds,
            self.c_active,
            self.c_enviro,
            self._c_loc_occ,
        ) = alpha

        self.dm_active = 2.0 * self.c_active @ self.c_active.T
        self.dm_enviro = 2.0 * self.c_enviro @ self.c_enviro.T

        self.beta_active_MO_inds = None
        self.beta_enviro_MO_inds = None
        self.beta_c_active = None
        self.beta_c_enviro = None
        self._beta_c_loc_occ = None
        self.beta_dm_active = None
        self.beta_dm_enviro = None

        if beta is not None:
            # Weight the DMs by 1 for unrestricted to combine them.
            self.dm_active *= 0.5
            self.dm_enviro *= 0.5

            (
                self.beta_active_MO_inds,
                self.beta_enviro_MO_inds,
                self.beta_c_active,
                self.beta_c_enviro,
                self._beta_c_loc_occ,
            ) = beta

            self.beta_dm_active = self.beta_c_active @ self.beta_c_active.T
            self.beta_dm_enviro = self.beta_c_enviro @ self.beta_c_enviro.T

        if sanity_check is True:
            self._check_values()

        if self._run_virtual_localization is True:
            logger.error("Virtual localization is not implemented.")
            # c_virtual = self._localize_virtual_orbs()
            # logger.error("Defualting to unlocalized virtual orbitals.")
            # c_virtual = self._global_ks.mo_coeff[:, self._global_ks.mo_occ < 2]
        else:
            logger.debug("Not localizing virtual orbitals.")
            # appends standard virtual orbitals from SCF calculation (NOT localized in any way)
            # c_virtual = self._global_ks.mo_coeff[:, self._global_ks.mo_occ < 2]

        # Unused
        # self.c_loc_occ_and_virt = np.hstack((self._c_loc_occ, c_virtual))

        logger.debug("Localization complete.")

In [8]:
import numpy as np
driver = NbedDriver(**args)

## Combine the C matrices and occupations into one to use restricted localizer
if driver._global_ks.mo_coeff.shape[0] == 2:
    driver._global_ks.mo_coeff = np.sum(driver._global_ks.mo_coeff, axis=0)/2
    driver._global_ks.mo_occ = np.sum(driver._global_ks.mo_occ, axis=0)
    
print(driver._global_ks.mo_coeff.shape)
old = OldSPADELocalizer(driver._global_ks, driver.n_active_atoms)
driver.localized_system = old
print(old.active_MO_inds)
print(old.enviro_MO_inds)

(35, 35)
[0 1 2 3 4]
[ 5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


In [9]:
from nbed.localizers import SPADELocalizer


class UnrestrictedDriver(NbedDriver):
    def __init__(
        self,
        geometry: str,
        n_active_atoms: int,
        basis: str,
        xc_functional: str,
        projector: str,
        localization: Optional[str] = "spade",
        convergence: Optional[float] = 1e-6,
        charge: Optional[int] = 0,
        spin: Optional[int] = 0,
        mu_level_shift: Optional[float] = 1e6,
        run_ccsd_emb: Optional[bool] = False,
        run_fci_emb: Optional[bool] = False,
        run_virtual_localization: Optional[bool] = False,
        run_dft_in_dft: Optional[bool] = False,
        max_ram_memory: Optional[int] = 4000,
        pyscf_print_level: int = 1,
        savefile: Optional[Path] = None,
        unit: Optional[str] = "angstrom",
        occupied_threshold: Optional[float] = 0.95,
        virtual_threshold: Optional[float] = 0.95,
        init_huzinaga_rhf_with_mu: bool = False,
        max_hf_cycles: int = 50,
        max_dft_cycles: int = 50,
    ):
        """Initialise class."""
        logger.debug("Initialising driver.")
        config_valid = True
        if projector not in ["mu", "huzinaga", "both"]:
            logger.error(
                "Invalid projector %s selected. Choose from 'mu' or 'huzinzaga'.",
                projector,
            )
            config_valid = False

        if localization not in ["spade", "ibo", "boys", "pipek-mezey"]:
            logger.error(
                "Invalid localization method %s. Choose from 'ibo','boys','pipek-mezey' or 'spade'.",
                localization,
            )
            config_valid = False

        if not config_valid:
            logger.error("Invalid config.")
            raise NbedConfigError("Invalid config.")

        self.geometry = geometry
        self.n_active_atoms = n_active_atoms
        self.basis = basis.lower()
        self.xc_functional = xc_functional.lower()
        self.projector = projector.lower()
        self.localization = localization.lower()
        self.convergence = convergence
        self.charge = charge
        self.spin = spin
        self.mu_level_shift = mu_level_shift
        self.run_ccsd_emb = run_ccsd_emb
        self.run_fci_emb = run_fci_emb
        self.run_virtual_localization = run_virtual_localization
        self.run_dft_in_dft = run_dft_in_dft
        self.max_ram_memory = max_ram_memory
        self.pyscf_print_level = pyscf_print_level
        self.savefile = savefile
        self.unit = unit
        self.occupied_threshold = occupied_threshold
        self.virtual_threshold = virtual_threshold
        self.max_hf_cycles = max_hf_cycles
        self.max_dft_cycles = max_dft_cycles

        self._check_active_atoms()
        self.localized_system = None
        self.two_e_cross = None
        self._dft_potential = None

        self._restricted_scf = False

        # self.embed(init_huzinaga_rhf_with_mu=init_huzinaga_rhf_with_mu) # TODO uncomment.
        logger.debug("Driver initialisation complete.")


## If the C matrix is combined before localisation, we get the right DFT components

In [90]:
driver = UnrestrictedDriver(**args)

## Combine the C matrices and occupations into one to use restricted localizer
if driver._restricted_scf is False:
    driver._global_ks.mo_coeff = np.sum(driver._global_ks.mo_coeff, axis=0)/2
    driver._global_ks.mo_occ = np.sum(driver._global_ks.mo_occ, axis=0)
    
print(driver._global_ks.mo_coeff.shape)
old = OldSPADELocalizer(driver._global_ks, driver.n_active_atoms)
driver.localized_system = old
print(old.active_MO_inds)
print(old.enviro_MO_inds)

(35, 35)
[0 1 2 3 4]
[ 5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


## If the density matrices are combined after localisation, it also passes!

In [126]:
driver = UnrestrictedDriver(**args)

## Combine the C matrices and occupations into one to use restricted localizer
print(driver._global_ks.mo_coeff.shape)
old = SPADELocalizer(driver._global_ks, driver.n_active_atoms)
driver.localized_system = old
print(old.active_MO_inds)
print(old.enviro_MO_inds)

driver.localized_system.dm_active += driver.localized_system.beta_dm_active
driver.localized_system.dm_enviro += driver.localized_system.beta_dm_enviro
driver.localized_system.beta_dm_active = np.zeros(driver.localized_system.dm_active.shape)
driver.localized_system.beta_dm_enviro = np.zeros(driver.localized_system.dm_enviro.shape)

(2, 35, 35)
occupancy=array([2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])
occupancy=array([2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])
[0 1 2 3 4]
[ 5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


In [127]:
"""Function to perform subsystem RKS DFT calculation."""
logger.debug("Calculating active and environment subsystem terms.")

def _ks_components(
    ks_system: Localizer,
    subsystem_dm: np.ndarray,
) -> Tuple[float, float, np.ndarray, np.ndarray, np.ndarray]:
    """Calculate the components of subsystem energy from a RKS DFT calculation.

    For a given density matrix this function returns the electronic energy, exchange correlation energy and
    J,K, V_xc matrices.

    Args:
        dm_matrix (np.ndarray): density matrix (to calculate all matrices from)

    Returns:
        Energy_elec (float): DFT energy defubed by input density matrix
        e_xc (float): exchange correlation energy defined by input density matrix
        J_mat (np.ndarray): J_matrix defined by input density matrix
    """
    logger.debug("Finding subsystem RKS componenets.")
    # It seems that PySCF lumps J and K in the J array
    two_e_term = ks_system.get_veff(dm=subsystem_dm)
    j_mat = two_e_term.vj
    # k_mat = np.zeros_like(j_mat)

    e_xc = two_e_term.exc
    # v_xc = two_e_term - j_mat

    energy_elec = (
        np.einsum("ij,ji->", ks_system.get_hcore(), subsystem_dm)
        + two_e_term.ecoul
        + two_e_term.exc
    )

    # if check_E_with_pyscf:
    #     energy_elec_pyscf = driver._global_ks.energy_elec(dm=dm_matrix)[0]
    #     if not np.isclose(energy_elec_pyscf, energy_elec):
    #         raise ValueError("Energy calculation incorrect")
    logger.debug(f"Subsystem RKS components found.")
    return energy_elec, e_xc, j_mat, two_e_term.ecoul

(driver.e_act, e_xc_act, j_act, e_coul_act) = _ks_components(
    driver._global_ks, driver.localized_system.dm_active
)
print(f'{driver.e_act=}, {e_xc_act=}, {e_coul_act=}')
(driver.e_env, e_xc_env, j_env, e_coul_env) = _ks_components(
    driver._global_ks, driver.localized_system.dm_enviro
)
print(f'{driver.e_env=}, {e_xc_env=}, {e_coul_env=}')
# print(f"{driver.e}")
# Computing cross subsystem terms
logger.debug("Calculating two electron cross subsystem energy.")

two_e_term_total = driver._global_ks.get_veff(
    dm=driver.localized_system.dm_active + driver.localized_system.dm_enviro
)
e_xc_total = two_e_term_total.exc

j_cross = 0.5 * (
    np.einsum("ij,ij", driver.localized_system.dm_active, j_env)
    + np.einsum("ij,ij", driver.localized_system.dm_enviro, j_act)
)
# Because of projection
k_cross = 0.0

xc_cross = e_xc_total - e_xc_act - e_xc_env
print(f"{xc_cross=}")

# overall two_electron cross energy
driver.two_e_cross = j_cross + k_cross + xc_cross
print(f"{e_xc_total=}, {j_cross=}, {k_cross=}, {xc_cross=}")


energy_DFT_components = (
    driver.e_act + driver.e_env + driver.two_e_cross + driver._global_ks.energy_nuc()
)
print("RKS components")
print(f"{driver.e_act=},{driver.e_env=},{driver.two_e_cross=},{driver._global_ks.energy_nuc()=}")
print(driver._global_ks.scf_summary)
print(f"{energy_DFT_components=}")
print(f"{driver._global_ks.e_tot=}")
if not np.isclose(energy_DFT_components, driver._global_ks.e_tot):
    logger.error(
        "DFT energy of localized components not matching supersystem DFT."
    )
    raise ValueError(
        "DFT energy of localized components not matching supersystem DFT."
    )


driver.e_act=-129.12067920087063, e_xc_act=-6.870575593286567, e_coul_act=32.699661799705716
driver.e_env=-329.78530535989285, e_xc_env=-23.580767460160047, e_coul_env=174.36536222278582
xc_cross=-0.34977914311295777
e_xc_total=-30.80112219655957, j_cross=76.06820092620217, k_cross=0.0, xc_cross=-0.34977914311295777
RKS components
driver.e_act=-129.12067920087063,driver.e_env=-329.78530535989285,driver.two_e_cross=75.71842178308921,driver._global_ks.energy_nuc()=189.1078142252856
{'e1': -635.5196655298082, 'coul': 283.13322494869374, 'exc': -30.801122197070047, 'nuc': 189.1078142252856}
energy_DFT_components=-194.0797485523887
driver._global_ks.e_tot=-194.07974855289893


# Localize separately

Create a driver to force unrestricted even for closed shells.

In [15]:
from nbed.localizers import SPADELocalizer
driver = NbedDriver(**args)
driver._restricted_scf = True
new = SPADELocalizer(driver._global_ks, driver.n_active_atoms)
print(new.active_MO_inds)
print(new.beta_active_MO_inds)
print(new.enviro_MO_inds)
print(new.beta_enviro_MO_inds)

occupancy=2.0


IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

# Mock Driver.embed()

In [ ]:
import logging
logger = logging.getLogger(__name__)
from nbed.localizers import Localizer
from typing import Optional, Tuple, Dict
import numpy as np
from pyscf import scf, dft, gto

init_huzinaga_rhf_with_mu = False

"""Generate embedded Hamiltonian.

Note run_mu_shift (bool) and run_huzinaga (bool) flags define which method to use (can be both)
This is done when object is initialized.
"""
logger.debug("Embedding molecule.")
localized_system = driver._localize()
driver.localized_system = localized_system

# logger.info(localized_system.active_MO_inds, localized_system.beta_active_MO_inds)
# logger.info(localized_system.enviro_MO_inds, localized_system.beta_enviro_MO_inds)

e_nuc = driver._global_ks.energy_nuc()

# Run subsystem DFT (calls localized rks)
#driver._subsystem_dft()


In [10]:
assert not driver._restricted_scf
driver.localized_system._global_ks.scf_summary

{'e1': -635.519665529809,
 'coul': 283.133224948694,
 'exc': -30.801122197070068,
 'nuc': 189.1078142252856}

In [11]:
both = [driver.localized_system.dm_active, driver.localized_system.beta_dm_active]
driver.localized_system._global_ks.get_j(dm=driver.localized_system.dm_active)

array([[ 6.12363740e+00,  1.20578053e+00,  4.39357600e-04, ...,
         1.92205480e-03,  1.65930871e-03,  7.57175983e-04],
       [ 1.20578053e+00,  2.97525652e+00,  4.07139523e-03, ...,
         1.76626871e-02,  2.71853140e-02,  1.59508992e-02],
       [ 4.39357600e-04,  4.07139523e-03,  2.97318136e+00, ...,
         2.55074822e-02,  3.74460150e-02,  2.19335118e-02],
       ...,
       [ 1.92205480e-03,  1.76626871e-02,  2.55074822e-02, ...,
         1.06638896e+00,  3.67881607e-01, -3.29573709e-01],
       [ 1.65930871e-03,  2.71853140e-02,  3.74460150e-02, ...,
         3.67881607e-01,  8.64592051e-01,  1.50143742e-01],
       [ 7.57175983e-04,  1.59508992e-02,  2.19335118e-02, ...,
        -3.29573709e-01,  1.50143742e-01,  8.13855497e-01]])

In [14]:
"""Function to perform subsystem RKS DFT calculation."""
logger.debug("Calculating active and environment subsystem terms.")

def _ks_components(
    ks_system: Localizer,
    subsystem_dm: np.ndarray,
    open_shell: bool,
) -> Tuple[float, float, np.ndarray, np.ndarray, np.ndarray]:
    """Calculate the components of subsystem energy from a RKS DFT calculation.

    For a given density matrix this function returns the electronic energy, exchange correlation energy and
    J,K, V_xc matrices.

    Args:
        dm_matrix (np.ndarray): density matrix (to calculate all matrices from)

    Returns:
        Energy_elec (float): DFT energy defubed by input density matrix
        e_xc (float): exchange correlation energy defined by input density matrix
        J_mat (np.ndarray): J_matrix defined by input density matrix
    """
    logger.debug("Finding subsystem RKS componenets.")
    # It seems that PySCF lumps J and K in the J array
    # need to access the potential for the right subsystem for unrestricted
    print(subsystem_dm.shape)
    two_e_term = ks_system.get_veff(dm=subsystem_dm)
    j_mat = ks_system.get_j(dm=subsystem_dm)
    k_mat = np.zeros_like(j_mat)

    # v_xc = two_e_term - j_mat

    if open_shell:
        j_tot = j_mat[0] + j_mat[1]
        dm_tot = subsystem_dm[0] + subsystem_dm[1]
    else:
        j_tot = j_mat
        dm_tot = subsystem_dm

    e_act = (
        np.einsum("ij,ji->", ks_system.get_hcore(), dm_tot)
        + 0.5*(np.einsum('ij,ji->', j_tot, dm_tot))
        + two_e_term.exc
    )

    # if check_E_with_pyscf:
    #     energy_elec_pyscf = driver._global_ks.energy_elec(dm=dm_matrix)[0]
    #     if not np.isclose(energy_elec_pyscf, energy_elec):
    #         raise ValueError("Energy calculation incorrect")
    logger.debug(f"Subsystem RKS components found.")
    return e_act, two_e_term, j_mat, k_mat

if not driver._restricted_scf:
    dm_act = np.array([driver.localized_system.dm_active, driver.localized_system.beta_dm_active])
    dm_env = np.array([driver.localized_system.dm_enviro, driver.localized_system.beta_dm_enviro])
else:
    dm_act = driver.localized_system.dm_active
    dm_env = driver.localized_system.dm_enviro

(e_act, two_e_act, j_act, k_act) = _ks_components(
    driver._global_ks, dm_act
)
# print(e_act, alpha_e_xc_act)
(e_env, two_e_env, j_env, k_env) = _ks_components(
    driver._global_ks, dm_env
)
# print(alpha_e_env, alpha_e_xc_env, alpha_ecoul_env)
driver.e_act = e_act
driver.e_env = e_env

# Computing cross subsystem terms
logger.debug("Calculating two electron cross subsystem energy.")
total_dm = driver.localized_system.dm_active + driver.localized_system.dm_enviro

if not driver._restricted_scf:
    total_dm += driver.localized_system.beta_dm_active + driver.localized_system.beta_dm_enviro
    
two_e_term_total = driver._global_ks.get_veff(
    dm=total_dm)
e_xc_total = two_e_term_total.exc

# if not driver._restricted_scf:
#     beta_two_e_term_total = driver._global_ks.get_veff(
#         dm=driver.localized_system.beta_dm_active + driver.localized_system.beta_dm_enviro
#     )
#     e_xc_total = (e_xc_total + beta_two_e_term_total.exc)/2


if driver._restricted_scf:
    j_cross = 0.5 * (
        np.einsum("ij,ij", driver.localized_system.dm_active, j_env)
        + np.einsum("ij,ij", driver.localized_system.dm_enviro, j_act)
        )
else:
    j_cross = 0.5 * (
        np.einsum("ij,ij", driver.localized_system.dm_active, j_env[0])
        + np.einsum("ij,ij", driver.localized_system.dm_enviro, j_act[0])
        + np.einsum("ij,ij", driver.localized_system.dm_active, j_env[1])
        + np.einsum("ij,ij", driver.localized_system.dm_enviro, j_act[1])
        + np.einsum("ij,ij", driver.localized_system.beta_dm_active, j_env[1])
        + np.einsum("ij,ij", driver.localized_system.beta_dm_enviro, j_act[1])
        + np.einsum("ij,ij", driver.localized_system.beta_dm_active, j_env[0])  
        + np.einsum("ij,ij", driver.localized_system.beta_dm_enviro, j_act[0])
    )

# Because of projection we expect kinetic term to be zero
k_cross = 0.0

xc_cross = e_xc_total - two_e_act.exc - two_e_env.exc

print(f'{xc_cross=}, {e_xc_total=}, {two_e_act.exc=}, {two_e_env.exc=}')
# if not driver._restricted_scf:
    # xc_cross = beta_e_xc_act - beta_e_xc_env
# driver.e_act=-129.12067009902808      driver.e_env=-329.78536772993164
# e_xc_total=-30.801124900358385, e_xc_act=-6.870585748333742, e_xc_env=-23.580761027571594
# j_cross=76.06825318343016 k_cross=0.0 xc_cross=-0.34977812445304934

# driver.e_act=-143.82987154686245 driver.e_env=-411.35820634557757
# e_xc_total=-30.80112219655959, alpha_e_xc_act=-2.6149703376981797, alpha_e_xc_env=-8.98550003899027, 
# j_cross=76.06820092620214, k_cross=0.0, xc_cross=-19.20065181987114

# overall two_electron cross energy
driver.two_e_cross = j_cross + k_cross + xc_cross
print(f"{e_xc_total=}, {j_cross=}, {k_cross=}, {xc_cross=}")

energy_DFT_components = (
    driver.e_act + driver.e_env + driver.two_e_cross + driver._global_ks.energy_nuc()
)
print("RKS components")
print(f'{driver.e_act=} {driver.e_env=} {driver.two_e_cross=}')
print(f'{driver._global_ks.scf_summary}')
print(f'{energy_DFT_components=}')
print(driver._global_ks.e_tot)
if not np.isclose(energy_DFT_components, driver._global_ks.e_tot):
    logger.error(
        "DFT energy of localized components not matching supersystem DFT."
    )
    raise ValueError(
        "DFT energy of localized components not matching supersystem DFT."
    )


(2, 35, 35)
(2, 35, 35)
xc_cross=-0.3497791425947874, e_xc_total=-30.801122196559604, two_e_act.exc=-6.87057559339729, two_e_env.exc=-23.580767460567525
e_xc_total=-30.801122196559604, j_cross=76.06820092620212, k_cross=0.0, xc_cross=-0.3497791425947874
RKS components
driver.e_act=-129.1206792009815 driver.e_env=-329.7853053603007 driver.two_e_cross=75.71842178360734
{'e1': -635.519665529809, 'coul': 283.133224948694, 'exc': -30.801122197070068, 'nuc': 189.1078142252856}
energy_DFT_components=-194.07974855238928
-194.07974855289945
CORRECT BITS:
e_xc_total, j_cross 
INCORRECT:
xc_cross : need to add cross terms for alpha on beta in each?


In [138]:
driver._global_ks.get_veff(dm=(driver.localized_system.dm_active + driver.localized_system.beta_dm_active)/2).exc + driver._global_ks.get_veff(dm=(driver.localized_system.dm_enviro + driver.localized_system.beta_dm_enviro)/2).exc

-11.600462187109128

driver.e_act=-129.12067920087074, e_xc_act=-6.870575593286571, e_coul_act=32.699661799705765

driver.e_env=-329.78530535989313, e_xc_env=-23.580767460160068, e_coul_env=174.36536222278605

xc_cross=-0.34977914311295066

e_xc_total=-30.80112219655957, j_cross=76.06820092620217, k_cross=0.0, xc_cross=-0.34977914311295777

RKS components

driver.e_act=-129.12067920087074,driver.e_env=-329.78530535989313,driver.two_e_cross=75.71842178308918,driver._global_ks.energy_nuc()=189.1078142252856

{'e1': -635.5196655298089, 'coul': 283.1332249486939, 'exc': -30.801122197070068, 'nuc': 189.1078142252856}
energy_DFT_components=-194.07974855238905
driver._global_ks.e_tot=-194.07974855289945

In [ ]:
"""Function to perform subsystem RKS DFT calculation."""
logger.debug("Calculating active and environment subsystem terms.")

def _rks_components(
    rks_system: Localizer, subsystem_dm: np.ndarray,
) -> Tuple[float, float, np.ndarray, np.ndarray, np.ndarray]:
    """Calculate the components of subsystem energy from a RKS DFT calculation.
    For a given density matrix this function returns the electronic energy, exchange correlation energy and
    J,K, V_xc matrices.
    Args:
        dm_matrix (np.ndarray): density matrix (to calculate all matrices from)
    Returns:
        Energy_elec (float): DFT energy defubed by input density matrix
        e_xc (float): exchange correlation energy defined by input density matrix
        J_mat (np.ndarray): J_matrix defined by input density matrix
    """
    dm_matrix = subsystem_dm
    # It seems that PySCF lumps J and K in the J array
    two_e_term = rks_system.get_veff(dm=dm_matrix)
    j_mat = two_e_term.vj
    # k_mat = np.zeros_like(j_mat)

    e_xc = two_e_term.exc
    # v_xc = two_e_term - j_mat

    energy_elec = (
        np.einsum("ij,ji->", rks_system.get_hcore(), dm_matrix)
        + two_e_term.ecoul
        + two_e_term.exc
    )

    # if check_E_with_pyscf:
    #     energy_elec_pyscf = driver._global_ks.energy_elec(dm=dm_matrix)[0]
    #     if not np.isclose(energy_elec_pyscf, energy_elec):
    #         raise ValueError("Energy calculation incorrect")

    return energy_elec, e_xc, j_mat

(driver.e_act, e_xc_act, j_act) = _rks_components(
    driver._global_ks, driver.localized_system.dm_active
)
(driver.e_env, e_xc_env, j_env) = _rks_components(
    driver._global_ks, driver.localized_system.dm_enviro
)
# Computing cross subsystem terms
logger.debug("Calculating two electron cross subsystem energy.")

two_e_term_total = driver._global_ks.get_veff(
    dm=driver.localized_system.dm_active + driver.localized_system.dm_enviro
)
e_xc_total = two_e_term_total.exc

j_cross = 0.5 * (
    np.einsum("ij,ij", driver.localized_system.dm_active, j_env)
    + np.einsum("ij,ij", driver.localized_system.dm_enviro, j_act)
)
# Because of projection
k_cross = 0.0

xc_cross = e_xc_total - e_xc_act - e_xc_env

# overall two_electron cross energy
driver.two_e_cross = j_cross + k_cross + xc_cross
print(f"{e_xc_total=}, {alpha_e_xc_act=}, {alpha_e_xc_env=}, {j_cross=}, {k_cross=}, {xc_cross=}")

energy_DFT_components = (
    driver.e_act + driver.e_env + driver.two_e_cross + driver._global_ks.energy_nuc()
)
logger.info("RKS components")
logger.info(driver.e_act)
logger.info(driver.e_env)
logger.info(driver.two_e_cross)
logger.info(driver._global_ks.energy_nuc())
print(energy_DFT_components, driver._global_ks.e_tot)
if not np.isclose(energy_DFT_components, driver._global_ks.e_tot):

    raise ValueError(
        "DFT energy of localized components not matching supersystem DFT"
    )

In [ ]:

logger.debug("Getting global DFT potential to optimize embedded calc in.")
g_act_and_env = driver._global_ks.get_veff(
    dm=(localized_system.dm_active + localized_system.dm_enviro)
)
g_act = driver._global_ks.get_veff(dm=localized_system.dm_active)
driver._dft_potential = g_act_and_env - g_act
logger.info(f"DFT potential average {np.mean(driver._dft_potential)}.")

# To add a projector, put it in this dict with a function
# if we want any more it's also time to turn them into a class
embeddings: Dict[str, callable] = {
    "huzinaga": driver._huzinaga_embed,
    "mu": driver._mu_embed,
}
if driver.projector in embeddings:
    embeddings = {driver.projector: embeddings[driver.projector]}

# This is reverse so that huz can be initialised with mu
for name in sorted(embeddings, reverse=True):
    logger.debug(f"Runnning embedding with {name} projector.")
    setattr(driver, "_" + name, {})
    result = getattr(driver, "_" + name)

    embedding_method = embeddings[name]
    local_rhf = driver._init_local_rhf()

    if init_huzinaga_rhf_with_mu and (name == "huzinaga"):
        logger.debug("Initializing huzinaga with mu-shift.")
        # seed huzinaga calc with mu result!
        result["v_emb"], result["scf"] = embedding_method(
            local_rhf, dmat_initial_guess=driver._mu["scf"].make_rdm1()
        )
    else:
        result["v_emb"], result["scf"] = embedding_method(local_rhf)

    result["mo_energies_emb_pre_del"] = local_rhf.mo_energy
    result["scf"] = driver._delete_environment(result["scf"], name)
    result["mo_energies_emb_post_del"] = local_rhf.mo_energy

    logger.info(f"V emb mean {name}: {np.mean(result['v_emb'])}")

    # calculate correction
    result["correction"] = np.einsum(
        "ij,ij", result["v_emb"], localized_system.dm_active
    )
    result["e_rhf"] = (
        result["scf"].e_tot
        + driver.e_env
        + driver.two_e_cross
        - result["correction"]
    )
    logger.info(f"RHF energy: {result['e_rhf']}")

    # classical energy
    result["classical_energy"] = (
        driver.e_env + driver.two_e_cross + e_nuc - result["correction"]
    )

    # Calculate ccsd or fci energy
    if driver.run_ccsd_emb is True:
        logger.debug("Performing CCSD-in-DFT embedding.")
        ccsd_emb, e_ccsd_corr = driver._run_emb_CCSD(
            result["scf"], frozen_orb_list=None
        )
        result["e_ccsd"] = (
            ccsd_emb.e_hf
            + e_ccsd_corr
            + driver.e_env
            + driver.two_e_cross
            - result["correction"]
        )
        logger.info(f"CCSD Energy {name}:\t{result['e_ccsd']}")

    if driver.run_fci_emb is True:
        logger.debug("Performing FCI-in-DFT embedding.")
        fci_emb = driver._run_emb_FCI(result["scf"], frozen_orb_list=None)
        result["e_fci"] = (
            (fci_emb.e_tot)
            + driver.e_env
            + driver.two_e_cross
            - result["correction"]
        )
        logger.info(f"FCI Energy {name}:\t{result['e_fci']}")

    if driver.run_dft_in_dft is True:
        did = driver.embed_dft_in_dft(driver._global_ks.xc, embedding_method)
        result["e_dft_in_dft"] = did["e_rks"]

if driver.projector == "both":
    logger.warning(
        "Outputting both mu and huzinaga embedding results as tuple."
    )
    driver.embedded_scf = (
        driver._mu["scf"],
        driver._huzinaga["scf"],
    )
    driver.classical_energy = (
        driver._mu["classical_energy"],
        driver._huzinaga["classical_energy"],
    )
elif driver.projector == "mu":
    driver.embedded_scf = driver._mu["scf"]
    driver.classical_energy = driver._mu["classical_energy"]
elif driver.projector == "huzinaga":
    driver.embedded_scf = driver._huzinaga["scf"]
    driver.classical_energy = driver._huzinaga["classical_energy"]

logger.info("Embedding complete.")